### ISSUE TITLE
`dotnet build` ignores Code Quality analysis rule severity settings in `.editorconfig` (VS honors them).

### ISSUE THREAD

**Issue by macrogreg** [dotnet/sdk#43574]
# Summary

`dotnet build` ignores Code Quality analysis rule severity settings in `.editorconfig`, while Visual Studio honors them.

# Environment

- Windows 10
- dotNet version: 8.0.304
- Visual Studio Community 2022 Version: 17.10.2

# Details / Repro

### 1. Configure Code Quality Analyzers with custom severity

I have a solution with several projects that represent some libraries and a gRPC-based backend service (Project Sdk="Microsoft.NET.Sdk.Web"). Everything targets and executes under .NET 8.

I have enabled Code Quality Analysis (CAXXXX-rules) by referencing the `Microsoft.CodeAnalysis.NetAnalyzers` package from a file named `CodeAnalysis.props`, which is in turn included from the `Directory.Build.props` located in the repository root.
(Note: the same exact issue is observed when I use the .NET SDK Analyzers via `EnableNETAnalyzers` (see later).)

> `Directory.Build.props`:
```xml
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <!-- ... -->     
    <Import Project="$(MSBuildThisFileDirectory)\Src\CodeAnalysis.props" Condition=" '$(SkipCodeAnalysis)' != 'True' " />    
</Project>
```

> `CodeAnalysis.props`:
```xml
<?xml version="1.0" encoding="utf-8"?>

<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">

	<!-- We use the code quality analyzer from the NuGet package (rather than the .NET SDK) to make sure that -->
	<!-- the analysis is identical regardless of the installed SDK version. Given the same NuGet package      -->
	<!-- and SDK versions the analysis should be identical. To switch to using the SDK analyzers, comment out -->
	<!-- the below package reference and change `EnableNETAnalyzers` to true.                                 -->	
	<PropertyGroup Condition=" '$(MSBuildProjectExtension)' == '.csproj' ">
		<EnableNETAnalyzers>false</EnableNETAnalyzers>
		<!--
		<EnableNETAnalyzers>true</EnableNETAnalyzers>
		-->
	</PropertyGroup>
	<!-- -->
	<ItemGroup Condition=" '$(MSBuildProjectExtension)' == '.csproj' ">
		<PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="8.0.0" />
	</ItemGroup>
	<!-- -->

	<!-- Detailed Code Quality Analysis Settings:                                                             -->
	<PropertyGroup Condition=" '$(MSBuildProjectExtension)' == '.csproj' ">
		<AnalysisMode>All</AnalysisMode>		
		<AnalysisLevel>8.0</AnalysisLevel>
		<_SkipUpgradeNetAnalyzersNuGetWarning>false</_SkipUpgradeNetAnalyzersNuGetWarning>
		<CodeAnalysisTreatWarningsAsErrors>false</CodeAnalysisTreatWarningsAsErrors>
	</PropertyGroup>
</Project>
```

Note that I set `<AnalysisMode>All</AnalysisMode>`.  
Next, I set up an `.editorconfig` file (in the same directory as `CodeAnalysis.props`). There, I configure custom severity for a subset of the rules:

> `.editorconfig`:
```ini
# Remove the line below if you want to inherit .editorconfig settings from higher directories
root = true

[*]
guidelines = 115 dotted 40404080

[*.cs]

# CA1024 [Design]: Use properties where appropriate
dotnet_diagnostic.CA1024.severity = none

# CA1031 [Design]: Do not catch general exception types
dotnet_diagnostic.CA1031.severity = silent

# CA1032 [Design]: Implement standard exception constructors
dotnet_diagnostic.CA1032.severity = none

# CA1034 [Design]: Nested types should not be visible
dotnet_diagnostic.CA1034.severity = none

# CA1303 [Globalization]: Do not pass literals as localized parameters
dotnet_diagnostic.CA1303.severity = none

# CA1513 [Maintainability]: Use ObjectDisposedException throw helper
dotnet_diagnostic.CA1513.severity = suggestion

# CA1805 [Performance]: Do not initialize unnecessarily
dotnet_diagnostic.CA1805.severity = none

# CA1812 [Performance]: Avoid uninstantiated internal classes
dotnet_diagnostic.CA1812.severity = warning

# CA1835 [Performance]: Prefer the memory-based overloads of ReadAsync/WriteAsync methods in stream-based classes
dotnet_diagnostic.CA1835.severity = none

# CA1848 [Performance]: Use the LoggerMessage delegates
dotnet_diagnostic.CA1848.severity = silent

# CA1852 [Performance]: Seal internal types
dotnet_diagnostic.CA1852.severity = none

# CA1859 [Performance]: Use concrete types when possible for improved performance
dotnet_diagnostic.CA1859.severity = none

# CA2000 [Reliability]: Dispose objects before losing scope
# (this would be such a great rule if it only would correctly skip over `Task` instances)
dotnet_diagnostic.CA2000.severity = silent

# CA2007 [Reliability]: Do not directly await a Task
# (most of the code is non a generic library running on ASP Core, so this is not applicable)
# (if we start creating generic libraries, we need to enable this rule for those libraries only)
dotnet_diagnostic.CA2007.severity = silent
```

### 2. Validate that the Code Analysis is working and everything builds as expected.

I build the solution (F6) and no issues are discovered.

![image](https://github.com/user-attachments/assets/969b6c91-ae2d-42dc-9f29-c3f20e5fe24f)

To verify that the Analyzers are working, I temporarily set in `.editorconfig`:
```ini
dotnet_diagnostic.CA2007.severity = error
```

Rebuilding the solution yields build-level errors as expected:

![image](https://github.com/user-attachments/assets/50fb30f3-e835-4821-af06-9a04923456e8)

The errors go away again after I restore the severity to silent.

### 3. Build from command line

Run `dotnet build --no-incremental` (using `dotnet build` does not make any difference).

The build output shows a large number of CAxxxx warnings that correspond to IDs that were disabled in the above `.editorconfig`.
In other words, the build system seems to respect the `<AnalysisMode>All</AnalysisMode>` setting and to completely ignore the overrides in `.editorconfig`.

### 4. Fixes tried

* Switching to the SDK analyzers results in the same issue (set `EnableNETAnalyzers` to true and remove the package reference in `CodeAnalysis.props`).

* Changing `AnalysisMode` to [another supported value](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#analysismode) seems to change the set of active rule accordingly, but any overrides in `.editorconfig` continue to be non-functional.

### 5. Impact and Question

The issue makes it impossible to enable the desired custom set of CA-rules in the CI.  

I cannot imagine I am the only person with this issue. Are there any known workarounds?

**Comment by macrogreg** [comment]
I have played around with various combinations of having VS open or not open while performing the command line builds.  
I appears that I can reproduce this about 3/4 of the time.
So it looks like this is related to some background process issue, probably the background analyzer in VS, although I find it surprising that the issue persists beyond closing of VS much, but not all of the time.

I have also determined that when run on a "clean" build agent, the issue does NOT occur, so the CI is no blocked.
With that in mind, I consider it not quite as critical.
Leaving it up here for now in case the bug report is useful to the .NET team.

**Comment by csteeg** [comment]
Any news on this? It's happening all the time for me :(

**Comment by rcocks-hl** [comment]
I get this happening "randomly".

Typically, suppressions are respected on the command-line, then sometimes it'll get into a state where running `dotnet build` fails for code analysis errors that should be suppressed. ( Usually ` SonarAnalyzer.CSharp` rules ).

Running `dotnet clean` tends to resolve the immediate issue for me, but a more permanent fix would be helpful.

**Comment by Keugon** [comment]
dotnet clean doesnt help any on my end, only solution is to disable warnings as errors the individual ovverride like:
dotnet_diagnostic.CA1040.severity = none,
dotnet_diagnostic.CA1998.severity = none
doesnt do anything here

**Comment by SeSamSeg** [comment]
Also ran into this. Created a minimal example [CA2007_build_twice_bug.zip](https://github.com/user-attachments/files/28049998/CA2007_build_twice_bug.zip). Created a duplicate issue on the wrong project.

If I run dotnet build on that project, the first run it gives the error, but downgraged to warning (but it shouldn't) and the second run it's no warning.

Inside of vs, it picks up that error the first run but when building the second time, it doesn't pick up the error as the previous build succeeded (cause this error got downgraded back to warning, while it was an error, not a warning)

I have the feeling that CodeAnalysisTreatWarningsAsErrors does not use the .editorconfig for restoring errors to warnings and only the rules defined by AnalysisMode and AnalysisLevel.

**Comment by SeSamSeg** [comment]
As a workaround I:
- don't set CodeAnalysisTreatWarningsAsErrors
- maintain a list of WarningsNotAsErrors with all warnings I don't want bumped to error
- (unsure if needed) copied the globalconfig from Sdks\Microsoft.NET.Sdk\analyzers\build\config\analysislevel_xxx to my .editorconfig

This makes the .editorconfig the single source of analysis config and allows layered config with solution + projects .editorconfig

The reason I want to keep MSBuildTreatWarningsAsErrors and WarningsNotAsErrors is that all msbuild and nuget warnings should def be looked at, also new ones in the future.

But as a fix, wouldn't adding _all_ code analysis rules to WarningsNotAsErrors when CodeAnalysisTreatWarningsAsErrors=true not be the most reliable way to implement CodeAnalysisTreatWarningsAsErrors?
